### PR TITLE
Cyclic dependency finder based on cmssw source (with or with header only dependencies)

### PR DESCRIPTION
**Pull request by davidlange6**
usage: cmssw_cycle_finder.py [-h] [--ignore_header_only] [--status_bar]

CMSSW Cyclic dependency finder

optional arguments:
  -h, --help            show this help message and exit
  --ignore_header_only  Ignore cycles due to header only dependencies
  --status_bar          Show progress bar when running

Requires python3.

### CODE REVIEW

**Comment by davidlange6** [comment]
please test

**Comment by davidlange6** [comment]
output on CMSSW_11_2_X_2020-09-08-2300

Building adjacency graph
Looking for cycles
Cycle detected: found a back edge from DataFormats/TrackerCommon to DataFormats/SiPixelDetId. It involves
DataFormats/TrackReco, DataFormats/SiPixelDetId, Alignment/CommonAlignment, TrackingTools/TrajectoryState, DataFormats/TrackerCommon, TrackingTools/TransientTrackingRecHit
Cycle detected: found a back edge from SimCalorimetry/HGCalSimProducers to SimCalorimetry/HGCalSimAlgos. It involves
SimCalorimetry/HGCalSimAlgos, SimCalorimetry/HGCalSimProducers

**Review by makortel — COMMENTED** [review]
The purpose is to find the cyclic dependencies between non-plugin libraries, right?

What about packages that define their `src` with `EDM_PLUGIN=1`? (I suppose including them does not harm, but thought to ask)

Would the plan be to run this in IBs (PRs?) in addition to ignominy, or eventually to replace it?

File: Utilities/ReleaseScripts/scripts/cmssw_cycle_finder.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+if __name__ == "__main__":
+    
+
+    include_header_only=True
+    show_status_bar=True
+
+    import argparse
+    parser=argparse.ArgumentParser(description="CMSSW Cyclic dependency finder")
+    parser.add_argument("--ignore_header_only",dest="include_header_only",
+                        action="store_true", default=False,
```
[the remark this answers is not in the document]
Comment (davidlange6) [reply Utilities/ReleaseScripts/scripts/cmssw_cycle_finder.py:151]: I'll fix this..

File: Utilities/ReleaseScripts/scripts/cmssw_cycle_finder.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    else:
+        return list(incs.keys()),list(pack_incs.keys())
+
+import sys
+
+if __name__ == "__main__":
+    
+
+    include_header_only=True
+    show_status_bar=True
```
[the remark this answers is not in the document]
Comment (davidlange6) [reply Utilities/ReleaseScripts/scripts/cmssw_cycle_finder.py:146]: no

File: Utilities/ReleaseScripts/scripts/cmssw_cycle_finder.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    for root, dirs, files in os.walk(pack_path, topdown=False):
+        for name in files:
+            for t in file_types:
+                if name.endswith(t):
+                    ret_val.append(os.path.join(root,name))
+
+    return ret_val
+    
+def get_sources(pack):
+    file_types = [".cc",".cpp"]
```
[the remark this answers is not in the document]
Comment (davidlange6) [reply Utilities/ReleaseScripts/scripts/cmssw_cycle_finder.py:98]: I added cxx - will need to look how to include CUDA (should be simple)

**Comment by davidlange6** [comment]
about - plugins - yes, src areas that are plugins are indeed included. as @dr15jones suggests these can create problems. I intend to also add plugins directories (but as separate entities), but have not done that yet.

**Comment by davidlange6** [comment]
I think this script also needs to use the set of -D used by CMSSW that ought to be included. that is unlikely to matter, so this is mostly a future note to myself..

File: Utilities/ReleaseScripts/scripts/cmssw_cycle_finder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidlange6) [reply Utilities/ReleaseScripts/scripts/cmssw_cycle_finder.py]: yes, supporting .cu was easy (we don't have any such files yet except in test)

**Comment by smuzaffar** [comment]
@davidlange6 , if you are only doing it using the CMSSW_RELEASE_BASE then why not make use of already created dependency information $CMSSW_RELEASE_BASE/etc/dependencies/uses.out.gz? Each line in this file is

src_file space_separated_list_of_all_included_headers

**Comment by davidlange6** [comment]
remind me how that list gets made? (I had forgotten about it even if I probably asked for it or helped create it..)

**Comment by smuzaffar** [comment]
We run this https://github.com/cms-sw/cmssw-config/blob/scramv3/SCRAM/findDependencies.py script for every release which basically parses the .d files generated by compiler. So one drawback/feature of this is that it only contains those files which are actually compiled and included. For example any header file which is not included by any source file will not end up in this.

**Comment by davidlange6** [comment]
Thanks!
Then indeed this can replace what I’ve done completely when the option of omitting header only dependencies is selected.

On Sep 11, 2020, at 6:03 PM, Malik Shahzad Muzaffar <notifications@github.com<mailto:notifications@github.com>> wrote:



We run this https://github.com/cms-sw/cmssw-config/blob/scramv3/SCRAM/findDependencies.py script for every release which basically parses the .d files generated by compiler. So one drawback/feature of this is that it only contains those files which are actually compiled and included. For example any header file which is not included by any source file will not end up in this.

**Comment by silviodonato** [comment]
@davidlange6 if I understand correctly, we close this PR. Is it correct?

**Comment by smuzaffar** [comment]
please test
@silviodonato , this is still a useful helper script. I would say lets get this in once tests passed (basically python3 test)

**Comment by smuzaffar** [comment]
+core

**Comment by qliphy** [comment]
+1

**Pull request merged**